### PR TITLE
added logging for slow RPC requests

### PR DESCRIPTION
**Pull request by dvovk**
Changed distribution of httpcfg.HttpCfg to be pointer.
Added new flags:
rpc.slow.log - which is false by default, this flag need to enable logging slow RPC requests
rpc.slow.log.threshold - which is 100 by default, this flag specify slow threshold in milliseconds 
Updated rpc handler to log slow requests:
- added map[request id] {method, timestamp}
- put every request details to map above
- delete request details from map above
- added time interval check for elements in map and if time difference is more than given threshold print request id and the method
- app will print slow requests in next cases:
1. As soon as request take more than given threshold
2. Every 20 seconds if request still in process
3. After request finished and it took more than give threshold

### CODE REVIEW

**Review by mh0lt — COMMENTED** [review]
I don't think you need both flags, you can just add the threshold and use  threshold!=0 as the flag.

**Review by mh0lt — COMMENTED** [review]
I don't think we should run this log message 10 times per second - it will be pretty expensive I'd have thought we want to run it say once every 20 secs ?

**Comment by dvovk** [comment]
> I don't think we should run this log message 10 times per second - it will be pretty expensive I'd have thought we want to run it say once every 20 secs ?
> 
> I think it makes the map logic a bit more complex. I think you should keep a start and end time and then remove it from the map once its logged.

I think frequency of log should depend on threshold value as all sense of this logs is to print slow requests which isn't finished yet. I agree that I can change logic to print slow request only on the first check

**Review by AskAlexSharov — CHANGES_REQUESTED** [review]

File: rpc/handler.go [hunk 1]
```diff
@@ -128,15 +139,49 @@ func newHandler(connCtx context.Context, conn jsonWriter, idgen func() ID, reg *
...
 
+func (h *handler) lookForSlowRpc() {
+	updateThreshold := 100 * time.Millisecond
+	slowThreshold := time.Duration(h.rpcSlowLogThreshold) * time.Millisecond
+	if slowThreshold < updateThreshold {
+		updateThreshold = slowThreshold
+	}
+
+	quit := make(chan struct{})
+	go func() {
```
[the remark this answers is not in the document]
Comment (dvovk) [reply rpc/handler.go:168]: if I do it it just hangs in the loop inside this coroutine and never process request

**Comment by dvovk** [comment]
> I don't think you need both flags, you can just add the threshold and use threshold!=0 as the flag.

Fixed

File: rpc/handler.go [hunk 2]
```diff
@@ -128,15 +139,49 @@ func newHandler(connCtx context.Context, conn jsonWriter, idgen func() ID, reg *
...
+		for {
+			select {
+			case <-ticker.C:
+				for k, v := range h.diagMessages {
+					if time.Since(v.Time) > slowThreshold {
+						h.logger.Warn("Slow RPC call", "method", v.Method, "reqid", idForLog{json.RawMessage(k)})
+					}
+				}
+			case <-quit:
+				ticker.Stop()
```
[the remark this answers is not in the document]
Comment (dvovk) [reply rpc/handler.go]: Fixed

File: rpc/handler.go [hunk 3]
```diff
@@ -128,15 +139,49 @@ func newHandler(connCtx context.Context, conn jsonWriter, idgen func() ID, reg *
...
 }
 
+func (h *handler) lookForSlowRpc() {
+	updateThreshold := 100 * time.Millisecond
+	slowThreshold := time.Duration(h.rpcSlowLogThreshold) * time.Millisecond
+	if slowThreshold < updateThreshold {
+		updateThreshold = slowThreshold
+	}
+
+	quit := make(chan struct{})
```
[the remark this answers is not in the document]
Comment (dvovk) [reply rpc/handler.go]: Fixed

**Comment by dvovk** [comment]
> > I don't think we should run this log message 10 times per second - it will be pretty expensive I'd have thought we want to run it say once every 20 secs ?
> > I think it makes the map logic a bit more complex. I think you should keep a start and end time and then remove it from the map once its logged.
> 
> I think frequency of log should depend on threshold value as all sense of this logs is to print slow requests which isn't finished yet. I agree that I can change logic to print slow request only on the first check

Fixed, added 20 second log interval

File: rpc/handler.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (AskAlexSharov) [reply rpc/handler.go:166]: Because you didn’t listen rootctx.Done()

File: rpc/handler.go [hunk 4]
```diff
@@ -128,15 +142,64 @@ func newHandler(connCtx context.Context, conn jsonWriter, idgen func() ID, reg *
 
 		maxBatchConcurrency: maxBatchConcurrency,
 		traceRequests:       traceRequests,
+
+		rpcSlowLogThreshold: rpcSlowLogThreshold,
+		diagMessage:         diagMsg{Method: "", Time: time.Unix(0, 0), LastLogTime: time.Unix(0, 0), Finished: true},
+		rpcMethodsWhitelist: []string{"eth_call", "eth_getBlock"},
```
Comment (AskAlexSharov) [inline rpc/handler.go:148]: blacklist

File: rpc/handler.go [hunk 5]
```diff
@@ -128,15 +142,64 @@ func newHandler(connCtx context.Context, conn jsonWriter, idgen func() ID, reg *
...
+
+	go func() {
+		ticker := time.NewTicker(updateThreshold)
+		defer ticker.Stop()
+
+		for {
+			select {
+			case <-ticker.C:
+				h.diagLock.Lock()
+				if h.isRpcMethodNeedsCheck(h.diagMessage.Method) {
```
Comment (AskAlexSharov) [inline rpc/handler.go:177]: `isRpcMethodNeedsCheck` check it once - before starting goroutine and ticker.

File: rpc/handler.go [hunk 6]
```diff
@@ -128,15 +142,64 @@ func newHandler(connCtx context.Context, conn jsonWriter, idgen func() ID, reg *
...
+func (h *handler) lookForSlowRpc() {
+	logInterval := 30 * time.Second
+	updateThreshold := 100 * time.Millisecond
+	slowThreshold := time.Duration(h.rpcSlowLogThreshold) * time.Millisecond
+	if slowThreshold < updateThreshold {
+		updateThreshold = slowThreshold
+	}
+
+	go func() {
+		ticker := time.NewTicker(updateThreshold)
```
Comment (AskAlexSharov) [inline rpc/handler.go:170]: why not `logInterval`?

File: rpc/handler.go [hunk 7]
```diff
@@ -128,15 +142,64 @@ func newHandler(connCtx context.Context, conn jsonWriter, idgen func() ID, reg *
...
+
+		for {
+			select {
+			case <-ticker.C:
+				h.diagLock.Lock()
+				if h.isRpcMethodNeedsCheck(h.diagMessage.Method) {
+					if !h.diagMessage.Finished {
+						requestDuration := time.Since(h.diagMessage.Time)
+						if requestDuration > slowThreshold && time.Since(h.diagMessage.LastLogTime) > logInterval {
+							h.logger.Warn("[Slow RPC] - request still processing", "method", h.diagMessage.Method, "Request duration", requestDuration)
```
Comment (AskAlexSharov) [inline rpc/handler.go:181]: let's use prefix `[rpc.slow]` - then it will be more grep-friendly

File: rpc/handler.go [hunk 8]
```diff
@@ -70,13 +70,26 @@ type handler struct {
 	serverSubs          map[ID]*Subscription
 	maxBatchConcurrency uint
 	traceRequests       bool
+
+	// metrics
+	diagMessage         diagMsg
+	diagLock            sync.Mutex
```
Comment (AskAlexSharov) [inline rpc/handler.go:76]: looks like we don't need this lock anymore

File: rpc/handler.go [hunk 9]
```diff
@@ -128,15 +142,64 @@ func newHandler(connCtx context.Context, conn jsonWriter, idgen func() ID, reg *
...
+	if rpcSlowLogThreshold > 0 {
+		h.lookForSlowRpc()
+	}
 	return h
 }
 
+func (h *handler) lookForSlowRpc() {
+	logInterval := 30 * time.Second
+	updateThreshold := 100 * time.Millisecond
+	slowThreshold := time.Duration(h.rpcSlowLogThreshold) * time.Millisecond
```
Comment (AskAlexSharov) [inline rpc/handler.go:164]: this can be done once - when parsing CLI flag.

**Review by AskAlexSharov — CHANGES_REQUESTED** [review]

**Comment by dvovk** [comment]
@AskAlexSharov  I rethink it a bit and removed a lot of stuff, I think now it is much better

File: rpc/handler.go [hunk 10]
```diff
@@ -388,7 +406,23 @@ func (h *handler) handleCallMsg(ctx *callProc, msg *jsonrpcMessage, stream *json
 		}
 		return nil
 	case msg.isCall():
+		if h.rpcSlowLogThreshold > 0 && h.isRpcMethodNeedsCheck(msg.Method) {
+			slowTimer := time.AfterFunc(time.Duration(h.rpcSlowLogThreshold)*time.Millisecond, func() {
```
Comment (AskAlexSharov) [inline rpc/handler.go:410]: really interesting. looks good for me

let's calc `time.Duration(h.rpcSlowLogThreshold)*time.Millisecond` once - when parsing cli flag.

File: rpc/handler.go [hunk 11]
```diff
@@ -388,7 +406,23 @@ func (h *handler) handleCallMsg(ctx *callProc, msg *jsonrpcMessage, stream *json
...
+				h.logger.Warn("[rpc.slow] - running", "method", msg.Method)
+			})
+
+			defer slowTimer.Stop()
+		}
+
 		resp := h.handleCall(ctx, msg, stream)
+
+		if h.rpcSlowLogThreshold > 0 && h.isRpcMethodNeedsCheck(msg.Method) {
```
Comment (AskAlexSharov) [inline rpc/handler.go:419]: let's store result of `h.isRpcMethodNeedsCheck(msg.Method)` into local bool variable

File: rpc/handler.go [hunk 12]
```diff
@@ -388,7 +406,23 @@ func (h *handler) handleCallMsg(ctx *callProc, msg *jsonrpcMessage, stream *json
 		}
 		return nil
 	case msg.isCall():
+		if h.rpcSlowLogThreshold > 0 && h.isRpcMethodNeedsCheck(msg.Method) {
+			slowTimer := time.AfterFunc(time.Duration(h.rpcSlowLogThreshold)*time.Millisecond, func() {
+				h.logger.Warn("[rpc.slow] - running", "method", msg.Method)
```
Comment (AskAlexSharov) [inline rpc/handler.go:411]: is it possible to print request ID and request body here? (and print only ID at `finish`)

**Review by AskAlexSharov — CHANGES_REQUESTED** [review]

**Comment by dvovk** [comment]
@AskAlexSharov changed mentioned stuff

**Comment by AskAlexSharov** [comment]
did push some changes - plz review

**Comment by dvovk** [comment]
> did push some changes - plz review

Looks ok

**Review by AskAlexSharov — APPROVED** [review]

**Pull request merged**